File: p062_euler.py
#
from datetime import datetime, timedelta

##################################################################
print("Solving projecteuler problem #062, Python solution #1")
start_time = datetime.now()
from itertools import count

cubes = {} 

def find(N):
    for n in count(start=345):
        key = ''.join(sorted(str(n*n*n)))
        if key in cubes:
            value = cubes[key]
            value[0] += 1
            if value[0] >= N:
                return value[1]**3
        else:
            cubes[key] = [1, n]

print(find(5))
dur_ms = ((datetime.now() - start_time).total_seconds())*1000
print('> duration in ms: {}'.format(dur_ms) )

##################################################################
import math
print("Solving projecteuler problem #062, Python solution #2")
start_time = datetime.now()
# parameters can be changed with these two variables
max_permutations = 5
upper_limit = 10000
number_code_dict = {}

for i in range(upper_limit):
    cube_value = int(math.pow(i, 3))

    # create a string of ordered digits for the dictionary key
    cube_value = "".join(sorted(str(cube_value)))

    if cube_value in number_code_dict:
        number_code_dict[cube_value].append(i)

        # test whether to print results and exit
        if len(number_code_dict[cube_value]) == max_permutations:
            print("Results:")
            for j in number_code_dict[cube_value]:
                print(j, "^3 =", int(math.pow(j, 3)))
            break
    else:
        # create new dictionary entry
        number_code_dict[cube_value] = [i]
dur_ms = ((datetime.now() - start_time).total_seconds())*1000
print('> duration in ms: {}'.format(dur_ms) )

##################################################################
print("Solving projecteuler problem #062, Python solution #3")
start_time = datetime.now()
main_dict = {}
sub_dict = {}
i = 2
while 5 not in main_dict.values():
    t = str(sorted(str(i ** 3)))
    if t in main_dict.keys():
        main_dict[t] += 1
        sub_dict[t].append(i)
    else:
        main_dict[t] = 1
        sub_dict[t] = [i]
    i += 1
for i in main_dict.keys():
    if main_dict[i] == 5:
        print(min([k**3 for k in sub_dict[i]]))        
dur_ms = ((datetime.now() - start_time).total_seconds())*1000
print('> duration in ms: {}'.format(dur_ms) )

File: p062_euler.js
/** https://projecteuler.net/problem=62
 * solution in Javascript
 * ian.leiman@gmail.com
 * 
 * The cube, 41063625 (345^3), can be permuted to produce two other cubes: 56623104 (384^3) and 66430125 (405^3). 
 * In fact, 41063625 is the smallest cube which has exactly three permutations of its digits which are also cube.
 * Find the smallest cube for which exactly five permutations of its digits are cube.
 * 
 * execute this on command line if you have node installed
 * node .\p062_euler.js
 * 
 * Answer:  127035954683
 */

/**
 * the first approach can be used to study cubes that have these permutations
 */
const answer = 127035954683;

function distSig(x){
    // calculates a digit distribution signature of a number
    // returns number of digits and array with how many digits each the number contains
    let xs = String(x);
    let sig = [0,0,0,0,0,0,0,0,0,0,0];
    siglen = xs.length; // set the length of the number in digits
    for (let i=xs.length-1; i >= 0; i--){ // check all the digits 
        digit = xs[i] ; // get the decimal digit
        sig[digit]++; // increment the count of this digit
    }
    // let sigstr = sig.reverse().join("");
    let sigstr = sig.join("");
    return [siglen, sigstr]; //parseInt( sigstr );
}

// configuration settings for this algorithm, can tweak to use for further 
// numerology studies
/*
the below settings will find all 12 digit cubes that have 5 permutations
    let printLimit = 3; // print only 5 permutation sets
    let maxRoot = 9999; // biggest root that has 12 digits
    let minRoot = 4642; // smallest root that has 12 digits
    let maxDig = 13; // stop before 13 digits 
    let minDig =12;  // start from 12 digits
*/
function p062_alg1(printLimit, maxRoot, minRoot, maxDig, minDig)
{
    // internal variables
    let root = 0;
    let cube = 0;
    let sig = []; 
    let sigbylen = [];
    let res = [];

    // init arrays for each digit lengths
    for (i= minDig; i < maxDig; i++){
        sigbylen[i] = Array();
    }

    for (root=minRoot; root < maxRoot; root++){
        cube = root*root*root;
        sig = distSig(cube);
        sigbylen[sig[0]].push( [root, cube, parseInt( sig[1] ) ] );
        // console.log(root, cube, sig);
    }

    for (i= minDig; i < maxDig; i++){
        let sl = sigbylen[i].length;
        console.log("***", i, sl);
        let sarr = sigbylen[i];
        let sigsorted = sarr.sort(function (a,b){return a[2] - b[2] });

        let key = 0;
        let count = 0;
        let sa = [];
        for(j=0; j < sl; j++){
            let thisSig = sigsorted[j][2];
            if (thisSig != key) {
                if (count > printLimit){
                    console.log( key, count+1, sa);
                    if (count == 4) {res.push(sa) }
                }
                key = thisSig;
                count = 0;
                sa = [];
            } else {
                if (count == 0) sa.push(sigsorted[j-1][1]); 
                count++; 
                sa.push(sigsorted[j][1]);  

            }
            //console.log( sigsorted[j] );

        }
    }
    let result = 9999999999999;
    for (i=0; i < res.length; i++){
        result = Math.min(result, res[i][0]);
    }
    correct = answer == result;
    console.log("#1 answer is: " + result + " is correct " + correct);
}


/**
 * another approach that is simpler and more brute force, slightly slower
 */
function p062_alg2()
{
    let oCube = [];
    for(var i=0; true; i++){
        var mCube = String(i*i*i);
        var cubeNum = mCube.split('');
        cubeNum.sort();
        var p = cubeNum.join('')
        if(oCube[p] == undefined){
            oCube[p] = [];
        }
        oCube[p].push(mCube);
        if(oCube[p].length>=5){
            console.log(oCube[p]);
            break;
        }
    }
    alg2result = oCube[p][0];
    correct = answer == alg2result;
    console.log("#2 answer is: " + alg2result + " is correct " + correct);
}

// here we call the different algorithms and time them
console.log("Solving projecteuler problem #062, solution #1");
console.time("1");
p062_alg1(3, 9999, 4642, 13, 12); // args can be tweaked to produce more or less output
console.timeEnd("1");

console.log("solution #2");
console.time("2");
p062_alg2();
console.timeEnd("2");
